codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/AuthSchemeSpecUtils.java

```diff
@@ -87,4 +92,55 @@ public boolean generateEndpointBasedParams() {
     public boolean includeParam(String name) {
         return allowedEndpointAuthSchemeParams.contains(name);
     }
+
+    public String serviceName() {
+        return intermediateModel.getMetadata().getServiceName();
+    }
+
+    public String signingName() {
+        return intermediateModel.getMetadata().getSigningName();
+    }
+
+    public AuthType defaultAuthType() {
+        return intermediateModel.getMetadata().getAuthType();
+    }
+
+    public Map<List<String>, List<AuthType>> operationsToAuthType() {
+        Map<List<AuthType>, List<String>> authSchemes2Operations =
+            intermediateModel.getOperations()
+                             .entrySet()
+                             .stream()
+                             .filter(kvp -> !kvp.getValue().getAuth().isEmpty())
+                             .collect(Collectors.groupingBy(kvp -> kvp.getValue().getAuth(),
+                                                            Collectors.mapping(Map.Entry::getKey, Collectors.toList())));
+
+        Map<List<String>, List<AuthType>> result = new HashMap<>();
+        authSchemes2Operations.forEach((key, value) -> result.put(value, key));
+
+        Set<String> operationsWithSpecificAuth = authSchemes2Operations
+            .values().stream()
+            .flatMap(List::stream)
+            .collect(Collectors.toSet());
+
+        // All the operations have a specific auth scheme configured, we can result here without caring about the service
+        // defaults.
+        if (operationsWithSpecificAuth.size() == intermediateModel.getOperations().size()) {
+            return result;
+        }
```

Comment:
would be good to have a test case for this too.

Reply:
Good catch, I found a bug there since the "default" is not added. I will remove this early optimization instead but still add a test cases for:

1. All operations with auth with the same values
2. All operations with auth with different values